sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs

```diff
@@ -0,0 +1,198 @@
+                        .Members
+                        .Select(issueFinder.FindIssue)
+                        .WhereNotNull()
+                        .ForEach(c.ReportDiagnostic);
+                },
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private class IssueFinder
+        {
+            private readonly IList<IMethodSymbol> allBaseClassMethods;
+            private readonly IList<IPropertySymbol> allBaseClassProperties;
+            private readonly SemanticModel semanticModel;
+
+            public IssueFinder(INamedTypeSymbol classSymbol, SemanticModel semanticModel)
+            {
+                this.semanticModel = semanticModel;
+                var allBaseClassMembers = GetAllBaseMembers(classSymbol, m => m.DeclaredAccessibility != Accessibility.Private);
+                allBaseClassMethods = allBaseClassMembers.OfType<IMethodSymbol>().ToList();
+                allBaseClassProperties = allBaseClassMembers.OfType<IPropertySymbol>().ToList();
+            }
+
+            private static IEnumerable<ISymbol> GetAllBaseMembers(INamedTypeSymbol classType, Func<ISymbol, bool> filter)
+            {
+                while (classType?.BaseType != null)
+                {
+                    foreach (var member in classType.BaseType.GetMembers().Where(filter))
+                    {
+                        yield return member;
+                    }
+
+                    classType = classType.BaseType;
+                }
+            }
+
+            public Diagnostic FindIssue(MemberDeclarationSyntax memberDeclaration)
+            {
+                var memberSymbol = semanticModel.GetDeclaredSymbol(memberDeclaration);
+
+                var methodSymbol = memberSymbol as IMethodSymbol;
+                if (methodSymbol != null)
+                {
+                    var hidingMethod = allBaseClassMethods.FirstOrDefault(
+                        m => DecrasesAccess(m.DeclaredAccessibility, methodSymbol.DeclaredAccessibility) &&
+                             IsMatchingSignature(m, methodSymbol));
+
+                    if (hidingMethod != null)
+                    {
+                        var location = (memberDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation();
+                        if (location != null)
+                        {
+                            return Diagnostic.Create(rule, location, hidingMethod);
+                        }
+                    }
+
+                    return null;
+                }
+
+                var propertySymbol = memberSymbol as IPropertySymbol;
+                if (propertySymbol != null)
+                {
+                    var hidingProperty = allBaseClassProperties.FirstOrDefault(p => DecreasesPropertyAccess(p, propertySymbol));
+                    if (hidingProperty != null)
+                    {
+                        var location = (memberDeclaration as PropertyDeclarationSyntax).Identifier.GetLocation();
```

Comment:
I would also extract this cast into a separate variable. Besides you use `as` cast but you are not checking for null here. Either it can't be null so you should use direct cast or it can and you should check for it.